multishard_mutation_query.cc

```diff
@@ -508,6 +504,23 @@ future<> read_context::save_reader(shard_id shard, const dht::decorated_key& las
                     reader->unpop_mutation_fragment(mutation_fragment_v2(schema, rparts->permit, *rit));
                 }
             }
+            if (rm.compaction_state) {
+                auto* next_fragment = co_await reader->peek();
+                // If there is nothing more left from the current partition --
+                // next fragment is either partition start, partition end or
```

Comment:
If the next fragment is a partition_end then unpopping is okay, yes? (as is not unpopping).

That leaves is with "next fragment is partition start on end of stream". Isn't that equivalent to "last fragment was partition end"? So we can change the condition to "if last fragment was a partition end, don't unpop the partition_start". This is nicer as it doesn't require the peek. I think the logic is more intuitive too - if we're in the middle of a partition, re-emit the partition start".

Reply:
If the next fragment is a partition-end, then pushing the partition header back is indeed not illegal, just pointless, because we'll just read back the empty partition on the next page.

Indeed checking if the last processed fragment was a partition-end is more accurate and simple, but very awkward to do as we use `flat_mutation_reader_v2::consume()` here and hence inspecting the fragment stream requires an interposer consumer. I'll try to check if there is a way to infer this inside the compactor (the consumer). It is not straightforward: the consume loop always calls `consume_partition_end()`, even if the consumption stopped in the middle of the partition (without actually seeing a partition-end fragment).

Reply:
Not that we very rarely have to actually wait here: we just pushed back a bunch of fragments just above, the reader will probably have fragments ready in its buffer, we just can't be 100% sure.

Reply:
I'm not worried about peek() cost, but about the extra logic and having more things than can go wrong.

For example, we'll read twice from an exhausted stream after this, whereas the standard practice is to read until end-of-stream and then stop.